### DISCUSSION TITLE
ARM64 Build

### DISCUSSION THREAD

**Discussion by ByronBoskellWork** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#1062] [Q&A]
I have been hoping to integrate this SDK with a Windows Universal Platform build. I have already had success creating a gstreamer pipeline on the platform and I was hoping to include the kvssink in the pipeline to stream camera audio / video similar to this example the first example pipeline in the repository.

I have been using Cmake GUI on a windows machine, I provided a toolchain file based off the example within in vcpkg included with visual studio. This successfully created a project in CMake, however when I try to compile to visual studio solution, I get the compile error: 

#error Atomics are not currently supported for non-x86 MSVC platforms.

Is there some additional setup that has to be done? I have found this issue:
https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/issues/827
Which implies that it is possible to compile on ARM64.

**Comment by niyatim23** [comment]
Hi @ByronBoskellWork, for a Windows build, please refer to the steps of a working Windows build in our CI:
https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/4c4afff63f3371a69be2472bb0bee654ef09abd3/.github/workflows/ci.yml#L278-L297

Be sure to modify the folder paths as per the location on your system. Additionally, you may also have to modify the bat file as per MSVC's path on your system and export the paths as follows:
```
set GST_PLUGIN_PATH=%CD%\build
set PATH=%PATH%;%CD%\open-source\local\bin;%CD%\open-source\local\lib
```

Check kvssink works with:
```
C:\Users\Administrator\Desktop\cpp\amazon-kinesis-video-streams-producer-sdk-cpp\build>gst-inspect-1.0 kvssink
Factory Details:
  Rank                     primary + 10 (266)
  Long-name                KVS Sink
  Klass                    Sink/Video/Network
  Description              GStreamer AWS KVS plugin
  Author                   AWS KVS <kinesis-video-support@amazon.com>

Plugin Details:
  Name                     kvssink
  Description              GStreamer AWS KVS plugin
  Filename                 C:\Users\Administrator\Desktop\cpp\amazon-kinesis-video-streams-producer-sdk-cpp\build\gstkvssink.dll
  Version                  1.0
  
```

Example GStreamer pipeline:

```
gst-launch-1.0 -v videotestsrc is-live=true ! video/x-raw,framerate=30/1,width=720,height=480 ! clockoverlay time-format="%a %B %d, %Y %I:%M:%S %p" ! openh264enc ! h264parse ! kvssink stream-name=demo-stream
```

**Reply by ByronBoskellWork** [reply]
I have successfully built the plugin for a x64 windows system, and successfully streamed the video from a webcam. The Windows Universal Platform is a separate build target with different requirements; however, the current issue is that the project is being generated with an atomics library which does not compile for ARM64.

**Comment by zygmuntszpak** [comment]
It appears that the issue may be related to the [aws-c-common](https://github.com/awslabs/aws-c-common/issues/1042), which currently does not support the ARM64 on Windows Universal Platform.  The remarks in the post say: 

> I would like to add that this support would be helpful for our team. Reason we need ARM64 support is for Hololens 2 devices (ARM64 Windows Architecture)

which is exactly the situation you are encountering. It looks like the aws-sdk-cpp also relies on the aws-c-common library, as indicated by the [following issue](https://github.com/aws/aws-sdk-cpp/discussions/2553). 

@niyatim23 Does this suggest that Kinesis currently simply cannot be made to work on the Windows Universal Platform with ARM64? If so, can you suggest some other workaround for integrating a  HoloLens   video/audio stream with an AWS service?

**Reply by niyatim23** [reply]
Hi @zygmuntszpak, the `amazon-kinesis-video-streams-producer-sdk-cpp` does not use aws-c-common as a dependency at all. So that seems unrelated. We have the build working in our CI with Windows intel. Do the same steps not work for your case? What build errors are you seeing?

**Reply by zygmuntszpak** [reply]
To follow up on what ByronBoskellWork posted below, the build on Windows Intel works fine, but as he states it is the combination of Windows  Universal Platform and ARM64, which poses the problem. Based on the error log he posted, it looks like the issue can be tracked down to the following file: 

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/57637ea593f4b43c509413a44d993ed08d7f2616/src/utils/src/AtomicsMsvc.h#L10C1-L12C7

The combination of Windows + ARM64 is not covered in your current CI, probably because it requires a self-hosted runner. See https://github.blog/changelog/2022-09-28-github-actions-self-hosted-runners-now-support-windows-arm-hardware/

**Reply by zygmuntszpak** [reply]
@niyatim23 Do you perhaps have any further suggestions for us? Is Windows Universal Platform and ARM64 not a supported platform?

**Reply by niyatim23** [reply]
@zygmuntszpak, that is not a supported platform. You can try to replicate the hardware setup that we have in our CI. The details for the same should be available [here](https://github.com/actions/runner-images/blob/main/images/win/Windows2022-Readme.md)

**Comment by ByronBoskellWork** [comment]
The Windows Universal Platform ([What's a Universal Windows Platform (UWP) app? - UWP applications | Microsoft Learn](https://learn.microsoft.com/en-us/windows/uwp/get-started/universal-application-platform-guide)) is a platform which is separate from a standard Windows OS. It has different APIs and constraints from a standard windows system. This is described in the linked Microsoft Learn article. An app which works for a standard windows system is not necessarily compatible with a Windows Universal Platform app and at minimum will have to recompiled to run on the windows universal platform. 

Arm64 is a different kind of chip architecture [ARM architecture family - Wikipedia](https://en.wikipedia.org/wiki/ARM_architecture_family) and has different constraints to a non-arm chip. 

The HoloLens is both ARM64 and Windows Universal Platform. This means that software compiled for an x64 Windows machine is not compatible with the HoloLens and must be recompiled for both the change in operating system and the change in chipset. 

I can confirm that the producer library builds and runs on a x64 Windows 10 machine. The first step in working with the SDK was to configure for my machine’s local environment and test it. Within this repository is the file: [amazon-kinesis-video-streams-producer-sdk-cpp/.github/workflows/ci.yml at develop · awslabs/amazon-kinesis-video-streams-producer-sdk-cpp](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/develop/.github/workflows/ci.yml) 

On line 278 is the set-up for running a build on a standard windows machine. As mentioned above this build is not compatible with the ARM64 Universal Platform build target. Further down is the Ubuntu ARM64 target. While that is an ARM64 build, a HoloLens cannot run Ubuntu, nor can utilities be freely installed so this not help either. 

To build for this platform set I have made use of the toolchain files provided by visual studio to cross-compile to the ARM64 UWP. 

When I do so it creates a project however it will not compile as I mentioned with the error: 

![ARM64Cmake](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/assets/96029214/e215de8f-8f80-438a-8cdc-8aea72f7facd)

Severity	Code	Description	Project	File	Line	Suppression State 

Error	C1189	#error:  Atomics are not currently supported for non-x86 MSVC platforms	kvspicUtils	C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\AtomicsMsvc.h	11
![VisualStudio](https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/assets/96029214/f61382a2-c97f-45fe-96b0-2b549820e959)

**Comment by ByronBoskellWork** [comment]
Build Log:
Build started...
1>------ Build started: Project: ZERO_CHECK, Configuration: Debug ARM64 ------
1>1>Checking Build System
2>------ Build started: Project: kvspicUtils, Configuration: Debug ARM64 ------
3>------ Build started: Project: kvspic, Configuration: Debug ARM64 ------
2>cl : command line  warning D9002: ignoring unknown option '-fPIC'
2>Allocators.c
3>cl : command line  warning D9002: ignoring unknown option '-fPIC'
3>AckParser.c
2>Atomics.c
3>AuthIntegration.c
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\AtomicsMsvc.h(11,1): fatal  error C1189: #error:  Atomics are not currently supported for non-x86 MSVC platforms
2>Base64.c
3>Callbacks.c
2>BitField.c
3>Client.c
2>BitReader.c
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\client\src\Client.c(115,1): warning C4102: 'CleanUp': unreferenced label
3>ClientEvent.c
2>Crc32.c
3>ClientState.c
2>Directory.c
3>FrameOrderCoordinator.c
2>DoubleLinkedList.c
3>InputValidator.c
2>DynamicLibrary.c
3>Stream.c
2>Endianness.c
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\client\src\Stream.c(1890,24): warning C4101: 'iterator': unreferenced local variable
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\client\src\Stream.c(1889,25): warning C4101: 'pExistingSerializedMetadata': unreferenced local variable
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\client\src\Stream.c(1891,12): warning C4101: 'data': unreferenced local variable
3>StreamEvent.c
2>ExponentialBackoffRetryStrategy.c
3>StreamState.c
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\ExponentialBackoffRetryStrategy.c(26,11): warning C4244: 'function': conversion from 'UINT64' to 'unsigned int', possible loss of data
2>FileIo.c
3>Duration.c
2>FileLogger.c
3>AivHeap.c
2>HashTable.c
3>Common.c
2>Hex.c
3>Heap.c
2>InstrumentedAllocators.c
3>HybridFileHeap.c
2>Logger.c
3>HybridHeap.c
2>Math.c
3>SystemHeap.c
2>Mutex.c
3>MkvGenerator.c
2>Semaphore.c
3>MkvStatics.c
2>Generating Code...
2>Compiling...
2>SingleLinkedList.c
3>Generating Code...
3>Compiling...
3>NalAdapter.c
2>StackQueue.c
3>SpsParser.c
2>String.c
3>State.c
2>Tags.c
2>Thread.c
3>TraceProfiler.c
2>Threadpool.c
3>Allocators.c
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(81,79): warning C4047: 'function': 'UINT64' differs in levels of indirection from 'PThreadData'
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(81,79): warning C4024: 'stackQueueRemoveItem': different types for formal and actual parameter 2
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(95,67): warning C4047: 'function': 'UINT64' differs in levels of indirection from 'PThreadData'
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(95,67): warning C4024: 'stackQueueRemoveItem': different types for formal and actual parameter 2
3>Atomics.c
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(301,61): warning C4047: 'function': 'PUINT64' differs in levels of indirection from 'PThreadData *'
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(301,61): warning C4024: 'stackQueueIteratorGetItem': different types for formal and actual parameter 2
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(309,67): warning C4047: 'function': 'UINT64' differs in levels of indirection from 'PThreadData'
2>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(309,67): warning C4024: 'stackQueueRemoveItem': different types for formal and actual parameter 2
2>ThreadsafeBlockingQueue.c
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\AtomicsMsvc.h(11,1): fatal  error C1189: #error:  Atomics are not currently supported for non-x86 MSVC platforms
2>Time.c
3>Base64.c
2>TimerQueue.c
3>BitField.c
2>Version.c
3>BitReader.c
2>Generating Code...
3>Crc32.c
3>Directory.c
2>Done building project "kvspicUtils.vcxproj" -- FAILED.
4>------ Build started: Project: kvsCommonCurl, Configuration: Debug ARM64 ------
3>DoubleLinkedList.c
4>Auth.c
3>DynamicLibrary.c
4>AwsV4Signer.c
3>Endianness.c
4>FileCredentialProvider.c
3>ExponentialBackoffRetryStrategy.c
4>IotCredentialProvider.c
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\ExponentialBackoffRetryStrategy.c(26,11): warning C4244: 'function': conversion from 'UINT64' to 'unsigned int', possible loss of data
3>FileIo.c
4>RequestInfo.c
3>FileLogger.c
3>HashTable.c
4>SslInit.c
3>Hex.c
4>StaticCredentialProvider.c
3>InstrumentedAllocators.c
4>Util.c
3>Generating Code...
3>Compiling...
3>Logger.c
4>Version.c
3>Math.c
4>jsmn.c
4>CurlCall.c
3>Mutex.c
4>CurlIotCredentialProvider.c
3>Semaphore.c
4>Generating Code...
3>SingleLinkedList.c
3>StackQueue.c
4>kvsCommonCurl.vcxproj -> C:\Users\ByronBoskell\aws-k-alt\UWPBuild\dependency\libkvscproducer\kvscproducer-src\Debug\kvsCommonCurl.lib
3>String.c
3>Tags.c
3>Thread.c
3>Threadpool.c
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(81,79): warning C4047: 'function': 'UINT64' differs in levels of indirection from 'PThreadData'
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(81,79): warning C4024: 'stackQueueRemoveItem': different types for formal and actual parameter 2
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(95,67): warning C4047: 'function': 'UINT64' differs in levels of indirection from 'PThreadData'
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(95,67): warning C4024: 'stackQueueRemoveItem': different types for formal and actual parameter 2
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(301,61): warning C4047: 'function': 'PUINT64' differs in levels of indirection from 'PThreadData *'
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(301,61): warning C4024: 'stackQueueIteratorGetItem': different types for formal and actual parameter 2
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(309,67): warning C4047: 'function': 'UINT64' differs in levels of indirection from 'PThreadData'
3>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\dependency\libkvspic\kvspic-src\src\utils\src\Threadpool.c(309,67): warning C4024: 'stackQueueRemoveItem': different types for formal and actual parameter 2
3>ThreadsafeBlockingQueue.c
3>Time.c
3>TimerQueue.c
3>Version.c
3>ContentView.c
3>Generating Code...
3>Done building project "kvspic.vcxproj" -- FAILED.
5>------ Build started: Project: cproducer, Configuration: Debug ARM64 ------
5>CallbacksProvider.c
5>ConnectionStaleStateMachine.c
5>ContinuousRetryStreamCallbacks.c
5>C:\Users\ByronBoskell\aws-k-alt\dependency\libkvscproducer\kvscproducer-src\src\source\ContinuousRetryStreamCallbacks.c(67,9): warning C4101: 'threadId': unreferenced local variable
5>CredentialProviderAuthCallbacks.c
5>CurlApiCallbacks.c
5>DeviceInfoProvider.c
5>FileAuthCallbacks.c
5>FileLoggerPlatformCallbackProvider.c
5>IotAuthCallback.c
5>Producer.c
5>Request.c
5>Response.c
5>StaticAuthCallbacks.c
5>StreamCallbacksProvider.c
5>StreamInfoProvider.c
5>StreamLatencyStateMachine.c
5>Generating Code...
5>cproducer.vcxproj -> C:\Users\ByronBoskell\aws-k-alt\UWPBuild\dependency\libkvscproducer\kvscproducer-src\Debug\cproducer.lib
5>Done building project "cproducer.vcxproj".
6>------ Build started: Project: KinesisVideoProducer, Configuration: Debug ARM64 ------
6>Building Custom Rule C:/Users/ByronBoskell/aws-k-alt/CMakeLists.txt
6>Auth.cpp
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(116,111): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(116,67): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(121,61): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(120,111): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(120,67): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(163,111): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(163,67): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(169,61): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(168,111): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>C:\Users\ByronBoskell\aws-k-alt\src\Auth.cpp(168,67): warning C4267: 'argument': conversion from 'size_t' to 'UINT32', possible loss of data
6>CachingEndpointOnlyCallbackProvider.cpp
6>CallbackProvider.cpp
6>DefaultCallbackProvider.cpp
6>DefaultDeviceInfoProvider.cpp
6>GetTime.cpp
6>KinesisVideoProducer.cpp
6>C:\Users\ByronBoskell\aws-k-alt\src\KinesisVideoProducerMetrics.h(63,31): warning C4244: 'return': conversion from 'const DOUBLE' to 'uint64_t', possible loss of data
6>KinesisVideoStream.cpp
6>C:\Users\ByronBoskell\aws-k-alt\src\KinesisVideoProducerMetrics.h(63,31): warning C4244: 'return': conversion from 'const DOUBLE' to 'uint64_t', possible loss of data
6>StreamDefinition.cpp
6>StreamTags.cpp
6>PutFrameHelper.cpp
6>C:\Users\ByronBoskell\aws-k-alt\src\KinesisVideoProducerMetrics.h(63,31): warning C4244: 'return': conversion from 'const DOUBLE' to 'uint64_t', possible loss of data
6>IotCertCredentialProvider.cpp
6>RotatingCredentialProvider.cpp
6>Generating Code...
6>KinesisVideoProducer.vcxproj -> C:\Users\ByronBoskell\aws-k-alt\UWPBuild\Debug\KinesisVideoProducer.lib
6>Done building project "KinesisVideoProducer.vcxproj".
7>------ Build started: Project: ALL_BUILD, Configuration: Debug ARM64 ------
7>Building Custom Rule C:/Users/ByronBoskell/aws-k-alt/CMakeLists.txt
========== Build: 5 succeeded, 2 failed, 0 up-to-date, 0 skipped ==========
========== Build started at 1:28 PM and took 21.747 seconds ==========

**Comment by ByronBoskellWork** [comment]
Toolchain file:

set(VCPKG_TARGET_ARCHITECTURE arm)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)

set(VCPKG_CMAKE_SYSTEM_NAME WindowsStore)
set(VCPKG_CMAKE_SYSTEM_VERSION 10.0)
